File: SoloLearn Codes/Word Rank/WordRank.cs
using System;
using System.Collections.Generic;

namespace SoloLearn
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(FindRank(Console.ReadLine()));
        }

        public static long FindRank(string word)
        {
            long rank = 1;
            long suffixPermCount = 1;
            Dictionary<char, int> charCounts = new Dictionary<char, int>();

            for (int i = word.Length - 1; i >= 0; i--)
            {
                char x = word[i];
                int xCount = charCounts.ContainsKey(x) ? charCounts[x] + 1 : 1;

                charCounts[x] = xCount;

                foreach (KeyValuePair<char, int> e in charCounts)
                {
                    if (e.Key < x)
                    {
                        rank += suffixPermCount * e.Value / xCount;
                    }
                }

                suffixPermCount *= word.Length - i;
                suffixPermCount /= xCount;
            }

            return rank;
        }
    }
}

File: SoloLearn Codes/Word Rank/WordRank.java
import java.util.*;

public class WordRank {

    public static void main(String[] args) {
        System.out.println(findRank(new Scanner(System.in).nextLine()));
    }
    
    public static long findRank(String word) {
        long rank = 1;
        long suffixPermCount = 1;
        final Map<Character, Integer> charCounts = new HashMap<>();

        for (int i = word.length() - 1; i >= 0; i--) {
            char x = word.charAt(i);
            int xCount = charCounts.containsKey(x) ? charCounts.get(x) + 1 : 1;

            charCounts.put(x, xCount);

            for (Map.Entry<Character, Integer> e : charCounts.entrySet()) {
                if (e.getKey() < x) {
                    rank += suffixPermCount * e.getValue() / xCount;
                }
            }

            suffixPermCount *= word.length() - i;
            suffixPermCount /= xCount;
        }

        return rank;
    }
}